classes/Cart.php

```diff
@@ -3888,26 +3888,47 @@ public function isVirtualCart()
         }
 
         if (!isset(self::$_isVirtualCart[$this->id])) {
-            $isVirtual = !$this->hasRealProducts();
+            if (!$this->hasProducts()) {
+                $isVirtual = false;
+            } else {
+                $isVirtual = !$this->hasRealProducts();
+            }
```

Comment:
if `hasRealProducts` calls `hasProducts`, we can probably revert this part of the PR?

Reply:
we can't revert this part, else it would break the previous behaviour that was available since PS 1.5 and broken into 1.7.4.0. In the previous version, `isVirtualCart` was returning false when cart was empty. If we only use `$isVirtual = !$this->hasRealProducts();` then `isVirtualCart` will return `true` in this case.

Reply:
Yep, we need to rely on the 2 methods, but when you look at the changes you applied in `hasRealProducts`, it calls `hasProducts` as well.

This means we call `hasProducts()` 2 times if we keep this part.

Reply:
You're right, i've removed the check into `hasRealProducts` method as it's only used right here.